core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -448,133 +448,42 @@ private[log] class Cleaner(val id: Int,
-
-        stats.readMessage(size)
-        if (shallowMessage.compressionCodec == NoCompressionCodec) {
-          if (shouldRetainMessage(source, map, retainDeletes, shallowMessageAndOffset, stats)) {
-            ByteBufferMessageSet.writeMessage(writeBuffer, shallowMessage, shallowOffset)
-            stats.recopyMessage(size)
-            if (shallowMessage.timestamp > maxTimestamp) {
-              maxTimestamp = shallowMessage.timestamp
-              offsetOfMaxTimestamp = shallowOffset
-            }
-          }
-          messagesRead += 1
-        } else {
-          // We use the absolute offset to decide whether to retain the message or not (this is handled by the
-          // deep iterator). Because of KAFKA-4298, we have to allow for the possibility that a previous version
-          // corrupted the log by writing a compressed message set with a wrapper magic value not matching the magic
-          // of the inner messages. This will be fixed as we recopy the messages to the destination segment.
-
-          var writeOriginalMessageSet = true
-          val retainedMessages = new mutable.ArrayBuffer[MessageAndOffset]
-          val shallowMagic = shallowMessage.magic
-
-          for (deepMessageAndOffset <- ByteBufferMessageSet.deepIterator(shallowMessageAndOffset)) {
-            messagesRead += 1
-            if (shouldRetainMessage(source, map, retainDeletes, deepMessageAndOffset, stats)) {
-              // Check for log corruption due to KAFKA-4298. If we find it, make sure that we overwrite
-              // the corrupted entry with correct data.
-              if (shallowMagic != deepMessageAndOffset.message.magic)
-                writeOriginalMessageSet = false
-
-              retainedMessages += deepMessageAndOffset
-              // We need the max timestamp and last offset for time index
-              if (deepMessageAndOffset.message.timestamp > maxTimestamp)
-                maxTimestamp = deepMessageAndOffset.message.timestamp
-            } else {
-              writeOriginalMessageSet = false
-            }
-          }
-          offsetOfMaxTimestamp = if (retainedMessages.nonEmpty) retainedMessages.last.offset else -1L
-          // There are no messages compacted out and no message format conversion, write the original message set back
-          if (writeOriginalMessageSet)
-            ByteBufferMessageSet.writeMessage(writeBuffer, shallowMessage, shallowOffset)
-          else if (retainedMessages.nonEmpty) {
-            val retainedSize = compressMessages(writeBuffer, shallowMessage.compressionCodec, retainedMessages)
-            stats.recopyMessage(retainedSize)
-          }
-        }
-      }
+      val result = messages.filterInto(writeBuffer, shouldRetain)
+
+      stats.readMessages(result.messagesRead, result.bytesRead)
+      stats.recopyMessages(result.messagesRetained, result.bytesRetained)
+
+      position += result.bytesRead
 
-      position += messages.validBytes
```

Comment:
Good question. We only increment `bytesRead` on message boundaries, so it seems correct.